File: count_digits.py
#!/usr/bin/env python3

"""This solves an old math brain teaser/challenge from when I was in college.
The point is to count every digit you would pass from
1 - 1,000,000. I have adopted the code to count to whatever you
like.  Anything past 1M will take quite some time.
Majority of code credit goes to Adam Watters.  My
version worked, but was clunky and took way too long.

Program originally written in Java which has my version of getting the correct answer.
"""

#intialize some variables for the program
#only doing 0s, 1s, and 2s. 2-9s should be the same.
onesW = 0
zeroesW =  0
twosW = 0

#get number to count to
gotonum = int(input("Give me the count number! "))
numLen = 0

#iterate through the number to count each digit

for i in range(1,gotonum+1):
  numLen = len(str(i))
  place = 10
  r = 0

  for j in range(1,numLen+1):
    if j == 1:
      r = i % 10

    else:
      r = (i//place) % 10
      place *= 10


    if r == 0:
      zeroesW += 1
    elif r == 1:
        onesW += 1
    elif r == 2:
        twosW += 1

print("Count Digits from 1 to " + str(gotonum))
print("0 = " + str(zeroesW))
print("1 = " + str(onesW))
print("2-9 = " + str(twosW))

File: CountDigits.java
/** Count each digit from 1 to whatever number you like...

	Not very efficient, but it works....good up until you hit the billions.


	The Second set of code at the bottom is more efficient thanks to Adam Watters!
  */

import java.io.*;
import java.util.*;


class CountDigits {
	
			
		
	public static int gotonum = 1000000;
	public static int numLen;

	public static void main(String[] args) {
		int ones = 0;
		int twos = 0;
		int threes = 0;
		int fours = 0;
		int fives = 0;
		int sixes = 0;
		int sevens = 0;
		int eights = 0;
		int nines = 0;    
		int zeroes = 0;
	
		System.out.println("\n\n1 to " + gotonum + "\n");
	
	// Initial loop for counting from 1 to whatever number....	
		for (int k = 1; k <= gotonum; k++) {
		   
		   // Turns the number to be counted into a string, gets the length, and puts it in an array.	
			numLen = Integer.toString(k).length();
			String digits = Integer.toString(k);
			char[] arr = digits.toCharArray();


		   //	System.out.println(arr);                       // Testing purposes
		   //	int i = Integer.parseInt(numLen);		// Testing purposes
			
		   // For loop to check each individual digit and compare below	
			for (int j = 0; j <= numLen-1; j++) {
				int i = Integer.parseInt(arr[j] + "");
			

				if (i==1) { ones++;}
				else if (i==2){twos++;}
			/*	else if (i==3){threes++;}
				else if (i==4){fours++;}
				else if (i==5){fives++;}
				else if (i==6){sixes++;}
				else if (i==7){sevens++;}
				else if (i==8){eights++;}
				else if (i==9){nines++;}     */
				else if (i==0){zeroes++;}
			}
			//System.out.println("X = " + i);
		}
		System.out.println("0 = " + zeroes);
		System.out.println("1 = " + ones);
		System.out.println("2-9 = " + twos);
	/*	System.out.println("3 = " + threes);
		System.out.println("4 = " + fours);
		System.out.println("5 = " + fives);
		System.out.println("6 = " + sixes);
		System.out.println("7 = " + sevens);
		System.out.println("8 = " + eights);
		System.out.println("9 = " + nines);  */

/****************************
	  ADAM WATTERS CODE BELOW.....TRANSLATED TO JAVA BY ME!
*****************************/

		System.out.println("\n" + "Watters Code now!");
  		
		int onesW = 0;
		int twosW = 0;
		int zeroesW = 0;
		
		for (int p = 1; p <= gotonum; p++) {
			numLen = Integer.toString(p).length();
			int place = 10;
			int r = 0;

			for ( int q = 1; q <= numLen; q++) {
				if (q == 1) { r = p % 10; }
				else  { r = (p/place)%10; 
					place *= 10;      }
				
				if (r==1) { onesW++;}
				else if (r==2){twosW++;}
				else if (r==0){zeroesW++;}

			}
						
		}
		System.out.println("0 = " + zeroesW);
		System.out.println("1 = " + onesW);
		System.out.println("2-9 = " + twosW);
	}
	
}

